lib/encoding.ts

```diff
@@ -270,3 +270,57 @@ class CompactedCMPRule implements AutoConsentCMPRule {
         return this.r[9].map(this._decodeRuleStep.bind(this));
     }
 }
+
+export function deduplicateRules(rules: AutoConsentCMPRule[]) {
+    const rulesBySelector = new Map<string, AutoConsentCMPRule[]>();
+    const dedupedRules: AutoConsentCMPRule[] = [];
+    const dedupKey = (rule: AutoConsentCMPRule) => {
+        if (rule.intermediate || !rule.runContext?.urlPattern) {
+            return null; // do not deduplicate intermediate or generic rules
+        }
+        return JSON.stringify({
+            mainFrame: rule.runContext?.main,
+            frame: rule.runContext?.frame,
+            detectCmp: rule.detectCmp,
+            detectPopup: rule.detectPopup,
+            optOut: rule.optOut,
+            prehideSelectors: rule.prehideSelectors,
+            cosmetic: rule.cosmetic,
+        });
+    };
+    rules.forEach((rule) => {
+        const key = dedupKey(rule);
+        if (key && typeof key === 'string') {
+            const patterns = rulesBySelector.get(key) || [];
+            patterns.push(rule);
+            rulesBySelector.set(key, patterns);
+        } else {
+            dedupedRules.push(rule);
+        }
+    });
+    rulesBySelector.forEach((rules, selector) => {
+        if (rules.length === 1) {
+            dedupedRules.push(rules[0]);
+            return;
+        }
+        dedupedRules.push({
+            name: `${rules[0].name}_+${rules.length - 1}`,
+            _metadata: {
+                deduplicatedFrom: rules.map((r) => r.name),
+            },
+            runContext: {
+                urlPattern: rules.map((rule) => rule.runContext!.urlPattern).join('|'),
+                main: rules[0].runContext?.main,
+                frame: rules[0].runContext?.frame,
+            },
+            prehideSelectors: rules[0].prehideSelectors,
+            detectCmp: rules[0].detectCmp,
+            detectPopup: rules[0].detectPopup,
+            optOut: rules[0].optOut,
+            optIn: [],
```

Comment:
It seems trivial to support optIn too, we just need to fill it here and add to `dedupKey`, right?